### PR TITLE
Make "calling set on destroyed object" error more descriptive

### LINKED ISSUE #10865: Make  "calling set on destroyed object" error more descriptive and helpful in debugging.
I wish to request Ember.js developers to consider changing error raised when set() is called on destroyed to include information that would help developers isolate faulty part of application and thus debug it more easily.

While I am yet to be biten by this error by my app, I already experienced plenty of day-long debugging sessions when my tests suite exploded with this error, and I was left in the dark about origins of problem.

Most basic solution for this issue would be to include name of property that was attempted to be set on destroyed object:

``Assertion Failed: tried to set 'PROPERTY' on destroyed object. This is usually caused by asynchronous code (i.e. Ember.run.later) attempting to change object state after it was destroyed.``

This change alone would be great improvement towards current situation. However I would like to take this step further, and thus make dev builds of Ember.js also `console.log` new value for property as well as object for which ``set`` failed.

### PR DESCRIPTION
**Pull request by duggiefresh**
Actual code is from PR #12264, I just added a test :)

Closes #10865

### CODE REVIEW

**Comment by duggiefresh** [comment]
There's a check here for `obj.constructor`, because at times `obj` will be the internal `EmptyObject`, and I don't know of a better way of checking for it. Otherwise, we receive a "can't turn an object into a primitive" error. ¯\_(ツ)_/¯

File: packages/ember-runtime/tests/system/object/observer_test.js [hunk 1]
```diff
@@ -103,7 +103,7 @@ testBoth('observer should not fire after being destroyed', function(get, set) {
 
   expectAssertion(function() {
     set(obj, 'bar', 'BAZ');
-  }, 'calling set on destroyed object');
+  }, 'calling set on destroyed object: <(unknown mixin):ember405>.bar = BAZ');
```
Comment (duggiefresh) [inline packages/ember-runtime/tests/system/object/observer_test.js]: Also, I'm not 100% sure this is okay to specify the ember id

File: packages/ember-metal/lib/property_set.js [hunk 2]
```diff
@@ -55,7 +55,8 @@ export function set(obj, keyName, value, tolerant) {
     return setPath(obj, keyName, value, tolerant);
   }
 
-  assert('calling set on destroyed object', !obj.isDestroyed);
+  assert(`calling set on destroyed object: ${obj.constructor ? obj : null}.${keyName} = ${value}`,
```
Comment (rwjblue) [inline packages/ember-metal/lib/property_set.js]: Does a `.toString` check satisfy the same edge cases with EmptyObject that this `obj.constructor` check does?  If so, I'd rather use it to make it clearer what we need.

`${obj.toString ? obj.toString() : ''}` or something?

**Comment by rwjblue** [comment]
Just had one tiny clarification, but I'm :+1:  on this...

**Comment by duggiefresh** [comment]
Yes that totally works! :tada: Thank you!

**Comment by rwjblue** [comment]
LGTM

**Comment by rwjblue** [comment]
Thanks @duggiefresh!

**Comment by bendemboski** [comment]
Apologies for my unclear commit comments. Here's what I'm seeing:

This PR introduces a regression where calling ```set()``` with an ```EmptyObject``` as the value throws a ```TypeError``` (on Chrome 48.0.2564.109 and I believe also on the latest Firefox). This is because ```EmptyObject``` is not convertible to a string, and the change to the set on destroyed object now unconditionally tries to convert the value to a string to construct the assertion message.

**Comment by mmun** [comment]
> This PR introduces a regression where calling `set()` with an `EmptyObject`...

How is this happening? EmptyObject is not public so it sounds like a framework bug?

**Comment by bendemboski** [comment]
I am seeing this triggered from Ember Data. An AdapterPopulatedRecordArray is setting a metadata object on itself, but the metadata object is an ```EmptyObject``` (https://github.com/emberjs/data/blob/master/addon/-private/system/empty-object.js).

**Comment by mmun** [comment]
I have a branch that adds a utils method to handle this case, but I haven't merged it in yet...

https://github.com/mmun/ember.js/commit/116b10b832d9aaec4219354f6611d6724d05b6e5

**Comment by bendemboski** [comment]
It's [this](https://github.com/emberjs/data/blob/v2.3.1/addon/-private/system/record-arrays/adapter-populated-record-array.js#L42) line in Ember Data 2.3.1 -- the store initializes its metadata objects to ```EmptyObject```, so that's meta's value in my repro. That code has since been changed a bit, so it might not occur there with later versions of Ember Data, but I think the underlying issue needs to be addressed in ```Ember.set``` regardless.

**Comment by mmun** [comment]
@bendemboski Thanks. That usage seems reasonable.

**Comment by mmun** [comment]
I'll PR a fix shortly.

**Comment by mmun** [comment]
I'll PR a fix shortly.

**Pull request merged**